pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_controller.go

```diff
@@ -0,0 +1,309 @@
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), obj.(*apisv1alpha1.APIExport))
+	logging.WithQueueKey(logger, key).V(2).Info("queuing APIExportEndpointSlices because of referenced APIExport")
+	for _, obj := range objectsForAPIExport {
+		slice, ok := obj.(*apisv1alpha1.APIExportEndpointSlice)
+		if !ok {
+			runtime.HandleError(fmt.Errorf("obj is supposed to be an APIExportEndpointSlice, but is %T", obj))
+			return
+		}
+		c.enqueueAPIExportEndpointSlice(slice)
+	}
+}
+
+// enqueueAllAPIExportEndpointSlices enqueues all APIExportEndpointSlices.
+func (c *controller) enqueueAllAPIExportEndpointSlices(shard interface{}) {
+	list, err := c.listAPIExportEndpointSlices()
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), shard.(*corev1alpha1.Shard))
+	for i := range list {
+		key, err := kcpcache.MetaClusterNamespaceKeyFunc(list[i])
+		if err != nil {
+			runtime.HandleError(err)
+			continue
+		}
+
+		logging.WithQueueKey(logger, key).V(2).Info("queuing APIExportEndpointSlice because Shard changed")
+		c.queue.Add(key)
+	}
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(4).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
+	clusterName, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return nil
+	}
+	obj, err := c.getAPIExportEndpointSlice(clusterName, name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+
+	old := obj
+	obj = obj.DeepCopy()
```

Comment:
do we need a copy here? A patch request should not modify an object, right?

Reply:
seems like we change both `old and obj` in https://github.com/kcp-dev/kcp/blob/8d95d652dd230a8b4a61598a9c7b77b32bf7972d/pkg/reconciler/committer/committer.go#L139

worth making sure we don't change object stored in the cache.